### ISSUE TITLE
Duplicates text when used with predictive text in iOS 17

### ISSUE THREAD

**Issue by ejbills** [kyle-n/HighlightedTextView#75]
**DO NOT DELETE THIS TEMPLATE**

If you do not fill out this template, your issue will be automatically closed. 

**Describe the bug**
When using this on iOS 17 with predictive text, the text duplicates itself. This makes typing within the editor impossible, disabling predictive text fixes this issue.

**To Reproduce**
Run it in an iOS 17 app environment and type in it.

**Expected behavior**
Normal functions, no duplication of text.

**Screenshots**

**Environment**
Please include:
- Exact version of iOS / iPadOS / macOS where your app was running when you saw the bug: iOS 17 (any version) iphone 15 pro
- Is your app using AppKit or UIKit? I am using it in SwiftUI.

**Additional context**

**Comment by kyle-n** [comment]
Thanks for reporting this. Seems like a big deal, since probably everyone leaves predictive text on. I'll look into it.

**Comment by kyle-n** [comment]
@ejbills I couldn't reproduce this issue.

This is the code I tried using on iOS 17.1.2. Predictive text did not cause any text to be duplicated.

```swift
struct MarkdownEditorA: View {
    @State var text: String = "Test"

    var body: some View {
        HighlightedTextEditor(text: $text, highlightRules: .markdown)
    }
}
```

https://github.com/kyle-n/HighlightedTextEditor/assets/13384477/4bb655d5-4f59-47b9-a97a-81c8a1d29935

Could you post a SwiftUI snippet that causes the issue, as well as what version of iOS your app is running on? Maybe a screen recording of the issue happening?

**Comment by kyle-n** [comment]
@ejbills Could you post a SwiftUI snippet that causes the issue, as well as what version of iOS your app is running on? Maybe a screen recording of the issue happening?

**Comment by Sam-Spencer** [comment]
@kyle-n I just started running into this issue on iOS 18.2 (although it is replicable in earlier versions). 

It appears to occur when typing quickly and there are multiple updates from the system to the textfield's range. Although I haven't been able to outline a set of steps that makes this easy to replicate (other than quickly typing ~100+ characters), it does happen consistently.

Based on logging I've added, I'll see a large sudden spike in `selectedTextRange` during some update cycles. It looks like this corresponds directly to a change in `markedTextRange`. In some cases this causes duplication of text, but in many instances it causes an `NSRange out of bounds` exception to throw. For example, in the crash below the logs show that `uiView.textStorage.length` is `36` in the call to `updateUIView` just prior to the crash. Then, in the call to `updateUIView` when the crash occurred, `uiView.textStorage.length` is suddenly `74`.

I'm attaching a video of the behavior (when the video "freezes" is when the crash occurs):

https://github.com/user-attachments/assets/565646b9-2965-44ac-a0fb-09cae69a9e3b

I'm also including a stack trace for this:
```
*** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[NSBigMutableString substringWithRange:]: Range {74, 2} out of bounds; string length 74'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001804b910c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x0000000180092da8 objc_exception_throw + 72
	2   Foundation                          0x0000000180f356c8 -[NSString substringFromIndex:] + 0
	3   UIKitCore                           0x0000000185eb9238 -[UITextInputController _selectedText] + 68
	4   UIKitCore                           0x00000001851e9d40 -[_UIKeyboardStateManager utlizeEuclidModelForDictationReplacement] + 132
	5   UIKitCore                           0x00000001851c57f4 -[_UIKeyboardStateManager prepareForSelectionChange] + 336
	6   UIKitCore                           0x00000001851c59ec -[_UIKeyboardStateManager selectionWillChange:] + 68
	7   UIKitCore                           0x0000000185eab358 __52-[UITextInputController _coordinateSelectionChange:]_block_invoke + 84
	8   UIFoundation                        0x0000000184c2b22c -[NSTextStorage coordinateReading:] + 48
	9   UIKitCore                           0x0000000185eab2e0 -[UITextInputController _coordinateSelectionChange:] + 116
	10  UIKitCore                           0x0000000185eaac8c -[UITextInputController setSelectedTextRange:] + 292
	11  UIKitCore                           0x0000000185eab0c8 -[UITextInputController setSelectedRange:afterDelta:appliedToRange:] + 104
	12  UIKitCore                           0x0000000185ed705c -[UITextView setSelectedRange:] + 60
	13  UIKit                               0x00000001cbfb917c -[UITextViewAccessibility setSelectedRange:] + 216
	14  [REDACTED_APP].debug.dylib          0x000000010982943c $s15highlighteditor21HighlightedTextEditorV12updateUIView_7contextySo10UITextViewC_7SwiftUI0F20RepresentableContextVyACGtF + 8584
	15  [REDACTED_APP].debug.dylib          0x000000010982aef0 $s15highlighteditor21HighlightedTextEditorV7SwiftUI19UIViewRepresentableAadEP06updateG0_7contexty0G4TypeQz_AD0gH7ContextVyxGtFTW + 60
	16  SwiftUI                             0x00000001d3a79998 $s7SwiftUI32PlatformViewRepresentableAdaptor33_19642D833A8FE469B137699ED1426762LLV06updateD8Provider_7contexty10UIViewTypeQz_AA0cdE7ContextVyADyxGGtF + 236
	17  SwiftUI                             0x00000001d39bddec $s7SwiftUI17PlatformViewChildV11updateValueyyFyyXEfU_yyXEfU3_yyXEfU_yyXEfU_yyXEfU_ + 112
	18  SwiftUI                             0x00000001d39bdd04 $s7SwiftUI17PlatformViewChildV11updateValueyyFyyXEfU_yyXEfU3_yyXEfU_yyXEfU_ + 344
	19  SwiftUICore                         0x00000001d46e4d04 $s7SwiftUI16ViewRendererHostPAAE21performExternalUpdateyyyyXEF + 260
	20  SwiftUI                             0x00000001d39c25c4 $s7SwiftUI17PlatformViewChildV11updateValueyyFyyXEfU_yyXEfU3_yyXEfU_yyXEfU_7performL_4workyyyXE_tAA0cD13RepresentableRzlFyyXEfU_TA + 64
	21  SwiftUICore                         0x00000001d422a554 $s7SwiftUI6UpdateO6ensureyxxyKXEKlFZxyKXEfU_ + 48
	22  SwiftUICore                         0x00000001d42295f0 $s7SwiftUI6UpdateO6ensureyxxyKXEKlFZ + 96
	23  SwiftUI                             0x00000001d39bdb00 $s7SwiftUI17PlatformViewChildV11updateValueyyFyyXEfU_yyXEfU3_yyXEfU_ + 444
	24  SwiftUI                             0x00000001d39c254c $s7SwiftUI17PlatformViewChildV11updateValueyyFyyXEfU_yyXEfU3_TA + 96
	25  SwiftUICore                         0x00000001d43ecfa0 $s7SwiftUI16_withObservation2dox5value_0D00D8TrackingV11_AccessListVSg06accessI0txyKXE_tKlFxSpyAJGKXEfU_ + 84
	26  SwiftUICore                         0x00000001d4080a18 $s7SwiftUI16_withObservation2dox5value_0D00D8TrackingV11_AccessListVSg06accessI0txyKXE_tKlFxSpyAJGKXEfU_TA + 24
	27  SwiftUICore                         0x00000001d4225198 $ss17withUnsafePointer2to_q0_x_q0_SPyxGq_YKXEtq_YKs5ErrorR_Ri_zRi_0_r1_lF + 72
	28  SwiftUICore                         0x00000001d43ec684 $s14AttributeGraph12StatefulRuleP7SwiftUIE15withObservation2doqd__qd__yKXE_tKlFTm + 544
	29  SwiftUICore                         0x00000001d43ec3dc $s14AttributeGraph12StatefulRuleP7SwiftUIE15withObservation2doqd__qd__yKXE_tKlF + 56
	30  SwiftUI                             0x00000001d39bcba8 $s7SwiftUI17PlatformViewChildV11updateValueyyFyyXEfU_ + 3044
	31  SwiftUICore                         0x00000001d406ee14 $s7SwiftUI8SignpostV13traceInterval6object__7closurexyXlSg_s12StaticStringVSays7CVarArg_pGyXKxyXEtlF + 344
	32  SwiftUI                             0x00000001d39bbe80 $s7SwiftUI17PlatformViewChildV11updateValueyyF + 164
	33  SwiftUI                             0x00000001d2c9360c $s14AttributeGraph0A0VyACyxGqd__c5ValueQyd__RszAA12StatefulRuleRd__lufcADSPyqd__GXEfU_ySv_So11AGAttributeatcyXEfU_ySv_AJtcfu_TA + 28
	34  AttributeGraph                      0x00000001beeaef54 _ZN2AG5Graph11UpdateStack6updateEv + 532
	35  AttributeGraph                      0x00000001beeaf7a4 _ZN2AG5Graph16update_attributeENS_4data3ptrINS_4NodeEEEj + 432
	36  AttributeGraph                      0x00000001beebd040 _ZN2AG8Subgraph6updateEj + 828
	37  SwiftUICore                         0x00000001d43d40c8 $s7SwiftUI9GraphHostC17flushTransactionsyyF + 840
	38  SwiftUI                             0x00000001d2effc3c $s7SwiftUI14_UIHostingViewC14_renderForTest8intervalySd_tFyAA9GraphHostCXEfU_ + 20
	39  SwiftUICore                         0x00000001d44d8e44 $s7SwiftUI17ViewGraphDelegatePAAE06updateD04bodyqd__qd__AA0D4HostCXE_tlFqd__AA0cD0CXEfU_TA + 20
	40  SwiftUICore                         0x00000001d46e23c0 $s7SwiftUI16ViewRendererHostPAAE06updateC5Graph4bodyqd__qd__AA0cG0CXE_tlF + 100
	41  SwiftUICore                         0x00000001d46e7208 $s7SwiftUI17ViewGraphDelegatePAAE06updateD04bodyqd__qd__AA0D4HostCXE_tlF + 56
	42  SwiftUI                             0x00000001d2effc0c $s7SwiftUI14_UIHostingViewC16beginTransactionyyFyycfU_yycfU_yyXEfU_ + 144
	43  SwiftUI                             0x00000001d2f0845c $s7SwiftUI14_UIHostingViewC16beginTransactionyyFyycfU_yycfU_yyXEfU_TA + 20
	44  SwiftUICore                         0x00000001d422a554 $s7SwiftUI6UpdateO6ensureyxxyKXEKlFZxyKXEfU_ + 48
	45  SwiftUICore                         0x00000001d42295f0 $s7SwiftUI6UpdateO6ensureyxxyKXEKlFZ + 96
	46  SwiftUI                             0x00000001d2f08438 $s7SwiftUI14_UIHostingViewC16beginTransactionyyFyycfU_yycfU_TA + 64
	47  SwiftUICore                         0x00000001d3ef65f8 $sIeg_ytIegr_TR + 20
	48  SwiftUICore                         0x00000001d467b880 $sSo9NSRunLoopC7SwiftUIE11addObserveryyyycFZySo05CFRunbF3RefaSg_So0gB8ActivityVSvSgtcfU_Tf4ddd_n + 136
	49  CoreFoundation                      0x000000018041c21c __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 32
	50  CoreFoundation                      0x0000000180416a80 __CFRunLoopDoObservers + 528
	51  CoreFoundation                      0x0000000180416ef0 __CFRunLoopRun + 896
	52  CoreFoundation                      0x00000001804166f4 CFRunLoopRunSpecific + 552
	53  GraphicsServices                    0x00000001905e5b10 GSEventRunModal + 160
	54  UIKitCore                           0x0000000185b319dc -[UIApplication _run] + 796
	55  UIKitCore                           0x0000000185b35bd4 UIApplicationMain + 124
	56  SwiftUI                             0x00000001d34b9040 $s7SwiftUI17KitRendererCommon33_ACC2C5639A7D76F611E170E831FCA491LLys5NeverOyXlXpFAESpySpys4Int8VGSgGXEfU_ + 164
	57  SwiftUI                             0x00000001d34b8d68 $s7SwiftUI6runAppys5NeverOxAA0D0RzlF + 84
	58  SwiftUI                             0x00000001d323d01c $s7SwiftUI3AppPAAE4mainyyFZ + 148
	59  [REDACTED_APP].debug.dylib          0x0000000108c6c544 $s5[REDACTED_APP][REDACTED_COMPANY]AppV5$mainyyFZ + 40
	60  [REDACTED_APP].debug.dylib          0x0000000108c6c9f0 __debug_main_executable_dylib_entry_point + 12
	61  dyld                                0x0000000100cb9410 start_sim + 20
	62  ???                                 0x0000000100dda274 0x0 + 4309492340
	63  ???                                 0xbd5f800000000000 0x0 + 13645766133444247552
)
libc++abi: terminating due to uncaught exception of type NSException
```

---

Would really appreciate any insight you might have into this. Also, happy to open a new issue, but this one seemed relevant so I figured I'd comment first.

**Comment by Sam-Spencer** [comment]
@kyle-n Update 🎊: I was able to capture a recording of the "duplicate" text behavior. I'm able to replicate this behavior by short-circuiting the `updateUIView` function when the view's `markedTextRange` is non-nil:


```swift
guard uiView.markedTextRange == nil else {
    uiView.isScrollEnabled = true
    context.coordinator.updatingUIView = false
    return
}

// Other stuff down here like:
uiView.selectedTextRange = context.coordinator.selectedTextRange
// ...
```

### Screen Recording
https://github.com/user-attachments/assets/eaf9aa40-13eb-4526-aecc-89fa6445c2c8

### Stack Trace
```
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'attempt to create backwards NSTextRange from 222 to 19'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001804b910c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x0000000180092da8 objc_exception_throw + 72
	2   Foundation                          0x0000000180e67c70 _userInfoForFileAndLine + 0
	3   UIFoundation                        0x0000000184c002c8 -[NSTextRange initWithLocation:endLocation:] + 464
	4   UIKitCore                           0x0000000185602db8 -[_UITextKit2LayoutController _rangeForCharacterRange:] + 184
	5   UIKitCore                           0x0000000185607e40 -[_UITextKit2LayoutController textRangeForCharacterRange:] + 44
	6   UIKitCore                           0x0000000185eb0b0c __82-[UITextInputController replaceRangeWithTextWithoutClosingTyping:replacementText:]_block_invoke + 100
	7   UIKitCore                           0x0000000185eb0bd4 -[UITextInputController _performWhileSuppressingDelegateNotifications:] + 40
	8   UIKitCore                           0x0000000185eb0a34 -[UITextInputController replaceRangeWithTextWithoutClosingTyping:replacementText:] + 324
	9   UIKitCore                           0x0000000185ed4ac0 -[UITextView replaceRangeWithTextWithoutClosingTyping:replacementText:] + 76
	10  UIKitCore                           0x00000001855b86c0 -[UIKBInputDelegateManager replaceRange:oldText:withText:forReplaceAction:] + 768
	11  UIKitCore                           0x00000001851e8298 -[_UIKeyboardStateManager applyAutocorrection:] + 1200
	12  UIKitCore                           0x00000001851d87c8 -[_UIKeyboardStateManager acceptAutocorrection:executionContextPassingTIKeyboardCandidate:] + 1276
	13  UIKitCore                           0x00000001851d6f54 -[_UIKeyboardStateManager acceptAutocorrectionForWordTerminator:executionContextPassingTIKeyboardCandidate:] + 436
	14  UIKitCore                           0x00000001851d66e4 -[_UIKeyboardStateManager addInputEvent:executionContext:] + 580
	15  UIKitCore                           0x00000001851d6328 __106-[_UIKeyboardStateManager addInputString:withFlags:withInputManagerHint:withInputSource:executionContext:]_block_invoke + 312
	16  UIKitCore                           0x00000001851d617c -[_UIKeyboardStateManager addInputString:withFlags:withInputManagerHint:withInputSource:executionContext:] + 984
	17  UIKitCore                           0x00000001851d43d4 -[_UIKeyboardStateManager handleStringInput:withFlags:withInputManagerHint:executionContext:] + 220
	18  UIKitCore                           0x00000001851ef7b0 -[_UIKeyboardStateManager handleKeyWithString:forKeyEvent:executionContext:] + 764
	19  UIKitCore                           0x00000001851b15a4 -[_UIKeyboardStateManager performOperations:withTextInputSource:] + 36
	20  UIKitCore                           0x00000001851ef180 -[_UIKeyboardStateManager handleKeyEvent:executionContext:] + 2628
	21  UIKitCore                           0x00000001851ee3f4 __42-[_UIKeyboardStateManager handleKeyEvent:]_block_invoke + 36
	22  UIKitCore                           0x00000001859589a8 -[UIKeyboardTaskEntry execute:] + 200
	23  UIKitCore                           0x000000018595731c -[UIKeyboardTaskQueue continueExecutionOnMainThread] + 336
	24  UIKitCore                           0x0000000185957800 -[UIKeyboardTaskQueue addTask:breadcrumb:] + 116
	25  UIKitCore                           0x00000001851ee36c -[_UIKeyboardStateManager handleKeyEvent:] + 404
	26  CoreFoundation                      0x00000001804bfce0 __invoking___ + 144
	27  CoreFoundation                      0x00000001804bcc54 -[NSInvocation invoke] + 276
	28  UIKitCore                           0x0000000185f45580 -[UIRepeatedAction invoke] + 172
	29  UIKitCore                           0x0000000185f45758 -[UIRepeatedAction _preInvocationTimerFire] + 52
	30  UIKitCore                           0x0000000185b4da50 -[UIApplication _handleKeyboardPressEvent:] + 2188
	31  UIKitCore                           0x0000000185b3a6b8 -[UIApplication pressesBegan:withEvent:] + 324
	32  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	33  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	34  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	35  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	36  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	37  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	38  SwiftUI                             0x00000001d2ca6738 $s7SwiftUI22UIKitKeyPressResponderC12pressesBegan_4withyShySo7UIPressCG_So14UIPressesEventCSgtFTm + 492
	39  SwiftUI                             0x00000001d2ca6860 $s7SwiftUI22UIKitKeyPressResponderC12pressesBegan_4withyShySo7UIPressCG_So14UIPressesEventCSgtFToTm + 152
	40  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	41  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	42  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	43  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	44  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	45  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	46  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	47  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	48  SwiftUI                             0x00000001d2ca6738 $s7SwiftUI22UIKitKeyPressResponderC12pressesBegan_4withyShySo7UIPressCG_So14UIPressesEventCSgtFTm + 492
	49  SwiftUI                             0x00000001d2ca6860 $s7SwiftUI22UIKitKeyPressResponderC12pressesBegan_4withyShySo7UIPressCG_So14UIPressesEventCSgtFToTm + 152
	50  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	51  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	52  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	53  UIKitCore                           0x000000018602c6f4 -[UIScrollView pressesBegan:withEvent:] + 144
	54  UIKitCore                           0x00000001852558bc __43-[UICollectionView pressesBegan:withEvent:]_block_invoke + 48
	55  UIKitCore                           0x0000000185255754 -[UICollectionView _handlePresses:keyHandler:defaultHandler:] + 404
	56  UIKitCore                           0x0000000185255860 -[UICollectionView pressesBegan:withEvent:] + 128
	57  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	58  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	59  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	60  SwiftUI                             0x00000001d2ca6738 $s7SwiftUI22UIKitKeyPressResponderC12pressesBegan_4withyShySo7UIPressCG_So14UIPressesEventCSgtFTm + 492
	61  SwiftUI                             0x00000001d2ca6860 $s7SwiftUI22UIKitKeyPressResponderC12pressesBegan_4withyShySo7UIPressCG_So14UIPressesEventCSgtFToTm + 152
	62  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	63  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	64  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	65  UIKitCore                           0x0000000185b5c65c forwardTouchMethod + 352
	66  UIKitCore                           0x000000018602c6f4 -[UIScrollView pressesBegan:withEvent:] + 144
	67  UIKitCore                           0x0000000185b6b0bc -[UIWindow _sendButtonsForEvent:] + 308
	68  UIKitCore                           0x0000000185b6b544 -[UIWindow sendEvent:] + 572
	69  UIKitCore                           0x0000000185b4b80c -[UIApplication sendEvent:] + 376
	70  UIKit                               0x00000001cbedd2fc -[UIApplicationAccessibility sendEvent:] + 108
	71  UIKitCore      
libc++abi: terminating due to uncaught exception of type NSException
```

**Comment by Sam-Spencer** [comment]
@kyle-n Okay, sorry for all the back-to-back comments, but I think I've come up with a potential solution. Still need to do more testing, but I've pushed up a few commits to [this fork](https://github.com/nenosinc/HighlightedTextEditor/commit/9c1564ce78f391d469d6153dc932ffe270ffecb3) which appear to resolve this. 

FWIW, in that same fork I've also added a new `Observable` version of `HightlightedTextEditor` that in my testing is more performant than the existing version using `Binding` (and may resolve #25). Curious to hear your thoughts on this.
